### DISCUSSION TITLE
Building doesn't seem to remove files that I deleted in the project from the _site directory.

### DISCUSSION THREAD

**Discussion by arpeasal** [11ty/eleventy#2294] [Q&A] [answered]
Building doesn't seem to remove files that I deleted in the project from the _site directory.  My start script is: "`eleventy --serve --port 5000`".

I have a directory `/blog/css/recursos/imagens` for css related images. (Note: in .eleveny.js the input is listed as "blog")  When I started working on this project today, the folder only had one file, sands.jpg 

I changed the name of that file to fundo_sand.jpg
I also added the file fundo_vitrine.gif and then changed its name to fundo_green.gif.

I then rebuilt the project, expecting the background of the site to be gone, because I didn't change the code in the css file.  But the background was NOT gone.  After doing some checking, I discovered that the `_site/css/recursos/imagens` directory was different from the `/blog/css/recursos/imagens` directory.

In the /blog version, there were only two files:
fundo_green.gif and fundo_sand.jpg

But in the _site version there were 4 files:
fundo_vitrine.gif, fundo_green.gif, sands.jpg and fundo_sand.jpg

I tried deleting the _site directory BEFORE I did the build and the new_site directory does NOT have the deleted files

It appears that when I rebuild without deleting the _site directory, eleventy just adds any new files, but doesn't remove any files that were removed from the project.

Am I doing something wrong?  Is there a way to make sure rebuilding removes obsolete files?

I had originally thought that I would just remove the _site directory from the project before every rebuild (Is there a way to do that with the build script?).  But I plan to use rsync to upload any changes to the site to my server.  I'm no expert on how rsync functions "under the hood", but I'm concerned that if I have a completely new _site directory, rsync won't know which files in the new _site directory are the same as those corresponding files on the server and it will upload the entire new contents of the _site directory.  I don't want to re-upload the entire site every time I add a new post.

**Comment by pdehaan** [comment, marked as answer]
Correct, Eleventy doesn't delete the `dir.output` folder before each build.
We have a few existing issues around that, for your reading pleasure:
- https://github.com/11ty/eleventy/issues/19
- https://github.com/11ty/eleventy/issues/744
- https://github.com/11ty/eleventy/issues/1090

Personally I either do a `prebuild` script in my package.json file (which automatically gets run before `npm run build` script) which runs <kbd>rm -rf www/</kbd>, which works for me on macOS. If you want to be fancier (and more cross-env compatible) you could also use the [**rimraf**](https://www.npmjs.com/package/rimraf) or [**del**](https://www.npmjs.com/package/del) modules in your .eleventy.js file. Not sure how that impacts [Eleventy incremental builds](https://www.11ty.dev/docs/usage/incremental/), probably quite poorly; but I personally don't use incremental builds so #yolo.

**Reply by arpeasal** [reply]
Thanks!
Two things.
First, I may try your solution with the prebuild script.  I use linux, so it should work the same as on macOS.  **Do you know if my concerns about updating the site on my server with rsync are valid if I completely remove _site before rebuilding?  That is, will rsync upload only the new versions or will it upload the entire _site directory contents since _site was comletely rebuilt?**

Second, in reading some of the articles you referenced, one commenter said:

> Given the responses, the following appears to be a decent cross-platform and safe solution:
> 
```
> const del = require('del');
> 
> module.exports = function (eleventyConfig) {
>      const dirToClean = '_site/*';
>      del(dirToClean);
> 
>      // rest of .eleventy.js config
> };
```
> However, it still seems quite odd that eliminating an addPassthroughCopy invocation will not result in the elimination of the associated subdirectory within the _site directory.

**Do you think this will work to scrub all the obsolete files from the _site directory?**

**Reply by pdehaan** [reply]
I don't have any real experience w/ rsync so I'm not sure. I usually manually deploy to sites.

As for the second question, that should work, but I'd strongly suggest using the "sync" APIs when cleaning folders. IIRC, rimraf and del are async by default so it might be chaos if some files are copied via `.addPassthroughCopy()` and then deleted based on subtle timing issues.
Should be as simple as using `del.sync(dirToClean)` or switching to `const del = require("del").sync;`

https://www.npmjs.com/package/del#delsyncpatterns-options
https://www.npmjs.com/package/rimraf#rimrafsync

**Reply by arpeasal** [reply]
Thanks for the help.  It'll probably be a couple of days till I can try this out.

**Reply by arpeasal** [reply]
OK, I tested the original suggestion (see below) to see if it worked for me.  It did.
```
> const del = require('del');
> 
> module.exports = function (eleventyConfig) {
>      const dirToClean = '_site/*';
>      del(dirToClean);
> 
>      // rest of .eleventy.js config
> };
```
I then tried one of your suggested changes, specifically `const del = require("del")**.sync**;` and that worked also, so I'll stick with that.

However, in my code I used single quotes where you used double quotes.
`require('del').sync` vs `require("del")`  Does that make a difference?

Thanks a million for your help.

**Reply by pdehaan** [reply]
Nope, no difference between single or double quotes

**Comment by arpeasal** [comment]
Thanks again!

**Comment by hozza** [comment]
_Coming from Jekyll, I didn't even consider that 11ty wouldn't clean the output `_site` directory before building. That was a surprise! :laughing:_ 

The solution chosen [here](https://github.com/11ty/eleventy/discussions/2294#discussioncomment-2469859) does not seem to work in the future (2025). There is now a [plugin by kentaroi](https://github.com/kentaroi/eleventy-plugin-clean) but it seems complicated (keeps track of, and deletes only 11ty generated files) and this function should really be [built into 11ty imo](https://github.com/11ty/eleventy/issues/19#issuecomment-1193363368).

I've updated the above full delete commands to completely remove the contents of `_site` before generating/building (works with `--serve` too). 

```js
const fullclean = require('del').deleteSync;

module.exports = function (eleventyConfig) {
      fullclean('_site/*');
 
      // rest of .eleventy.js config
 };
```
I renamed `del` to `fullclean` for verbosity, updated the require to the new synchronous function name, and you must `npm install del` beforehand.

Edit: I've updated the code from `clean` to `fullclean` to avoid confusion with more detailed approach for https://github.com/11ty/eleventy/issues/19.

**Reply by Ryuno-Ki** [reply]
As @pdehaan mentioned before: https://github.com/11ty/eleventy/issues/19

Your approach would violate:

> Must not delete any files not generated by the tool.

**Reply by hozza** [reply]
I know, my requirements are different. 

I reference the tracking deletion ticket for people who want that nuanced functionality. This discussion is about a simpler full delete setup (but was non-functional due to updates in the `del` package).

> There is now a [plugin by kentaroi](https://github.com/kentaroi/eleventy-plugin-clean) but it seems complicated (keeps track of, and deletes only 11ty generated files...
> I've updated the above full delete commands to completely remove the contents of _site before generating/building...

Edit: I've updated the code from `clean` to `fullclean` to avoid confusion with more detailed approach for #19.

**Reply by Ryuno-Ki** [reply]
That should help the next person coming across this discussion. Thank you :orange_heart:

**Reply by hozza** [reply]
FYI: an ESM compatible config version would replace the `const fullclean = require('del').deleteSync;` require with `import { deleteSync as fullclean } from 'del';`
